src/ZoneServer/Database/ZoneDb.cs

```diff
@@ -1205,5 +1207,113 @@ private void LoadQuests(Character character)
 				}
 			}
 		}
+			
+
+		/// <summary>
+		/// Saves the characters's collections to the database.
+		/// Note that collections are account-level
+		/// </summary>
+		/// <param name="character"></param>
+		/// <exception cref="InvalidOperationException"></exception>
+		private void SaveCollections(Character character)
+		{
+			using (var conn = this.GetConnection())
+			using (var trans = conn.BeginTransaction())
+			{
+				using (var cmd = new MySqlCommand("DELETE FROM `collections` WHERE `accountId` = @accountId", conn, trans))
+				{
+					cmd.AddParameter("@accountId", character.AccountId);
+					cmd.ExecuteNonQuery();
+				}
+
+				foreach (var collectionId in character.Collections.GetList())
+				{
+					var collectionProgress = character.Collections.GetProgress(collectionId);
+
+					if (collectionProgress.Count > 0)
+					{
+						foreach (var collectionItem in collectionProgress)
+						{
+							using (var cmd = new InsertCommand("INSERT INTO `collections` {0}", conn, trans))
+							{
+								cmd.Set("accountId", character.AccountId);
+								cmd.Set("collectionId", collectionId);
+								cmd.Set("itemId", collectionItem);
+
+								cmd.Execute();
+							}
+						}
+					}
+					else
+					{
+						// we insert a collection with a blank item if you have the collection
+						// but haven't registered anything
+
+						using (var cmd = new InsertCommand("INSERT INTO `collections` {0}", conn, trans))
+						{
+							cmd.Set("accountId", character.AccountId);
+							cmd.Set("collectionId", collectionId);
+							cmd.Set("itemId", -1);
+
+							cmd.Execute();
+						}
+					}
+				}
+
+				trans.Commit();
+			}
+		}
+
+		/// <summary>
+		/// Loads the character's collections from the database.
+		/// </summary>
+		/// <param name="character"></param>
+		private void LoadCollections(Character character)
+		{
+			using (var conn = this.GetConnection())
+			{
+				var previousCollection = -1;
+				var currentItemList = new List<int>();
+
+				using (var cmd = new MySqlCommand("SELECT collectionid, itemid FROM `collections` WHERE `accountid` = @accountId ORDER BY collectionid", conn))
+				{
+					cmd.AddParameter("@accountId", character.AccountId);
+
+					using (var reader = cmd.ExecuteReader())
+					{
+						while (reader.Read())
+						{
+							var collectionId = reader.GetInt32("collectionId");
+							var itemId = reader.GetInt32("itemId");
+
+							if (collectionId != previousCollection)
+							{
+								// we're starting a new collection, save the previous one
+
+								if (previousCollection != -1)
+								{
+									character.Collections.AddSilent(previousCollection, currentItemList);
```

Comment:
Wouldn't it be easier to add the items one by one instead of building up a list?

Reply:
The loading code would definitely be simpler though, and hence easier to write and read^^ But you do have a point, even though locking in this context is entirely inconsequential, because the lock isn't being contented.

But I also just realized that you're storing the list that you create here inside the collection component... That seems risky to me, since you never know what a user might do with that. While _your_ code is fine, you're giving users access to the internals of the component. Someone might pass in a list that they then keep using for other things, which may have unintended effects. Someone might also pass a list to `AddSilent` with the expectation that the items in the list are _added_ to the collection, when in reality that call will overwrite whatever was there before--something that isn't documented at all. I would highly recommend changing it to add the contents of the list to the internal list.